### PR TITLE
AC-568-add-acknowledge-date-to-filter-out-previous-acknowledged-errors-monai

### PR DESCRIPTION
**Pull request by lillie-dae**
Signed-off-by: Lillie Dae <lillie.dae@answerdigital.com>

### Description

This should be a new endpoint to return failed workflows, that have an empty value of

acknowledgedWorkflowErrors: null

at the workflow level. 

The filter should return all failed workflows since the date. If the date format is invalid a 400 bad request should be returned.

ENDPOINT:

workflowsinstances/failed?acknowledged=2022-02-21

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] All tests passed locally.
- [ ] [Documentation comments](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments) included/updated.

### CODE REVIEW

File: src/WorkflowManager/WorkflowManager/Controllers/WorkflowInstanceController.cs [hunk 1]
```diff
@@ -139,14 +141,57 @@ public async Task<IActionResult> GetByIdAsync([FromRoute] string id)
...
+        [Route("failed")]
+        [HttpGet]
+        public async Task<IActionResult> GetFailedAsync([FromQuery] string acknowledged)
+        {
+            if (string.IsNullOrWhiteSpace(acknowledged))
+            {
+                return Problem($"Failed to validate, no {nameof(acknowledged)} parameter provided", $"{ENDPOINT}failed", BadRequest);
+            }
+
+            var parseResult = DateTime.TryParse(acknowledged, out var acknowledgedDateTime);
```
Comment (mocsharp) [inline src/WorkflowManager/WorkflowManager/Controllers/WorkflowInstanceController.cs:169]: Should we use an [IFormatProvider](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.tryparse?view=net-7.0#system-datetime-tryparse(system-string-system-iformatprovider-system-globalization-datetimestyles-system-datetime@)) here? Would this work for other date formats in different cultures?

File: tests/IntegrationTests/WorkflowExecutor.IntegrationTests/StepDefinitions/WorkflowInstanceStepDefintions.cs [hunk 2]
```diff
@@ -79,6 +82,45 @@ public void GivenIHaveWorkflowInstances(int count)
...
         }
 
+        [Given(@"I have (.*) failed workflow Instances with acknowledged workflow errors with mid date as (.*)")]
+        public void GivenIHaveWorkflowInstancesWithAcknowledgedWorkflowErrors(int count, string midDate)
+        {
+            _outputHelper.WriteLine($"Retrieving {count} workflow instances");
+            var listOfWorkflowInstance = new List<WorkflowInstance>();
+
+            var currentCulture = Thread.CurrentThread.CurrentCulture;
+            Thread.CurrentThread.CurrentCulture = new CultureInfo("en-GB");
```
Comment (mocsharp) [inline tests/IntegrationTests/WorkflowExecutor.IntegrationTests/StepDefinitions/WorkflowInstanceStepDefintions.cs:92]: Do we want to test different cultures?

Comment (lillie-dae) [reply tests/IntegrationTests/WorkflowExecutor.IntegrationTests/StepDefinitions/WorkflowInstanceStepDefintions.cs]: Removed this reliance on CurrentCulture in integration tests but also added culture checks in unit tests.

I do you think this resolves the issue?

File: src/WorkflowManager/WorkflowManager/Controllers/WorkflowInstanceController.cs (hunk 1 above)
Comment (lillie-dae) [reply src/WorkflowManager/WorkflowManager/Controllers/WorkflowInstanceController.cs]: This bit of code has been refactored so ASP.net handles parsing and in params has a bit of text specifiy it should be passed in as ISO 8601 standard YYYY-MM-DD, even YYYY-MM-DDT00:00 will work.

**Review by neildsouth — APPROVED** [review]

**Pull request merged**